### ISSUE TITLE
Consider a local package cache option instead of global cache

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/pub#1976]
Related to https://github.com/dart-lang/pub/issues/1959

The idea here is to have at least an opt in option where all fetched packages would use a local cache under `.dart_tool/pub/` instead of the global cache.

Advantages:

- Can clean up old deps on pub upgrade - avoids the issue where your global pub cache grows infinitely over time.
- It is safer to modify dependencies in the cache, such modifications only affect the current package not all packages. It is very easy today to accidentally modify the shared version of a package in the cache.

Disadvantages:

- Can't share versions of dependencies across all packages on a machine.

In the long term the local package cache should always consume less disk space overall, since the global cache continues to grow over time. I am skeptical how much re-use most users actually get in practice from the global cache anyways.

**Comment by zoechi** [comment]
Is there any chance this might turn into requests for having some dependencies in global and some in local cache? 
I think `pub global activate` always needs to be in global cache, right?
Are there other potential candidates? How is this handled in npm?

I think the growing cache is quite a minor disadvantage. 
The big advantage is that the current solution is really simple.

> It is very easy today to accidentally modify the shared version

Doing that is quite ugly anyway. Is there really a need to support that?

**Comment by jakemac53** [comment]
> I think pub global activate always needs to be in global cache, right?

Yes, that would always use the global cache. I don't think we can drop the global cache entirely but we should be able to make it a lot smaller.

> The big advantage is that the current solution is really simple.

There isn't really a difference in complexity imo - conceptually its the same. The only difference is where the cache is located, and then the minor step of cleaning up old dependencies which should be relatively straightforward.

> Doing that is quite ugly anyway. Is there really a need to support that?

It isn't really something we want to explicitly support no - but this strategy at least makes it much easier to fix if/when you accidentally do it, and it makes the effects be localized to your current package.

The reality today is that it is very easy to accidentally edit these files at which point your whole cache is corrupted. 

> Is there any chance this might turn into requests for having some dependencies in global and some in local cache?

We definitely wouldn't want to support that, it would be all or nothing.

**Comment by zoechi** [comment]
> I am skeptical how much re-use most users actually get in practice

I thought at first how many projects to I maintain concurrently and they are actually quite few.
When I work on a project it takes usually several months and then there are a few on the side.

But application projects are split into several packages and organized as monorepo.
Currently I'm working on a project that is split into 6 packages that share quite a few dependencies and they are dependencies of each other which drag in their dependencies transitively.

How should that be handles. Stick with global cache? have a local cache for monorepo? Accept that each package has a copy of the dependencies downloaded?

No strong opinion, just brainstorming.

**Comment by zoechi** [comment]
@matanlurey that sounds interesting. 
(I haven't checked if the `mono_repo` package provides anything in that direction - not using id)

**Comment by natebosch** [comment]
Another reason to be very cautious about this - `pub` isn't the only codebase that knows and cares about the pub cache:

https://github.com/dart-lang/sdk/blob/5424289f95de4876b55fe32d95e9b6c8a002dbdc/pkg/analyzer_cli/lib/src/analyzer_impl.dart#L251
https://github.com/dart-lang/dart-vim-plugin/blob/940addeedc7fd81cdc1d5765472c8f01a0d487d7/autoload/dart.vim#L166

**Comment by sigurdm** [comment]
Is this not supported already by doing something like `PUB_CACHE=.dart_tool/pub/cache pub get`?

**Comment by sigurdm** [comment]
I don't think we'll do more here. Using `PUB_CACHE` should work well enough. Please reopen or comment if we need more.
